### PR TITLE
Add GPU,CPU,RAM,VRAM stats to tensorboard

### PR DESCRIPTION
**Pull request by jodusan**
## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/tune/logger.py [hunk 1]
```diff
@@ -216,6 +253,7 @@ def __init__(self,
             self._logger_cls_list = loggers
         self._sync_function = sync_function
         self._log_syncer = None
+        self.monitor = UtilMonitor(0.7)
```
Comment (stefanpantic) [inline python/ray/tune/logger.py]: Make the monitor optional

File: python/ray/tune/logger.py [hunk 2]
```diff
@@ -119,6 +124,38 @@ def to_tf_values(result, path):
...
+        self.start()
+
+    def read_utilization(self):
+        self.values.setdefault("perf/cpu", []).append(float(psutil.cpu_percent(interval=None)))
+        self.values.setdefault("perf/ram", []).append(float(getattr(psutil.virtual_memory(), 'percent')))
+        for gpu in GPUtil.getGPUs():
+            self.values.setdefault("perf/gpu"+str(gpu.id), []).append(float(gpu.load))
+            self.values.setdefault("perf/vram"+str(gpu.id), []).append(float(gpu.memoryUtil))
+
+    def get_data(self):
```
Comment (pimpke) [inline python/ray/tune/logger.py:142]: I think you need a lock when entering get_data and read_utilization methods, to ensure thread-safety.

File: python/ray/tune/logger.py [hunk 3]
```diff
@@ -119,6 +124,38 @@ def to_tf_values(result, path):
...
+        ret_values = copy.deepcopy(self.values)
+
+        for key, val in self.values.items():
+            val.clear()
```
Comment (pimpke) [inline python/ray/tune/logger.py]: Why do you need to clear the values?

File: python/ray/tune/logger.py [hunk 4]
```diff
@@ -10,6 +10,11 @@
 import yaml
 import distutils.version
 import numbers
+import GPUtil
+from threading import Thread
+import time
+import psutil
+import copy
```
Comment (pimpke) [inline python/ray/tune/logger.py]: Sort imports.

File: python/ray/tune/logger.py (hunk 3 above)
Comment (jodusan) [reply python/ray/tune/logger.py]: avoiding double mean? (tensorboard does it)

File: python/ray/tune/logger.py (hunk 2 above)
Comment (jodusan) [reply python/ray/tune/logger.py:142]: threads in python are not really threads?

Comment (nikola-j) [reply python/ray/tune/logger.py:142]: https://stackoverflow.com/a/40072999

**Review by nikola-j — DISMISSED** [review]

**Review by stefanpantic — DISMISSED** [review]

File: python/ray/tune/logger.py [hunk 5]
```diff
@@ -119,6 +122,41 @@ def to_tf_values(result, path):
...
+    def __init__(self, delay):
+        super(UtilMonitor, self).__init__()
+        self.stopped = False
+        self.delay = delay  # Time between calls to GPUtil
+        self.values = defaultdict(list)
+        self.lock = threading.Lock()
+        self.start()
+
+    def read_utilization(self):
+        self.lock.acquire()
```
Comment (pimpke) [inline python/ray/tune/logger.py]: Use `with self.lock:` block instead of `acquire` and `release`.

Comment (jodusan) [reply python/ray/tune/logger.py]: nice touch!

**Review by pimpke — APPROVED** [review]

**Pull request merged**